DOCSP-28558: connection pool faq + reuse client tip (#636)
mongodb/docs-node: source/faq.txt

pool removes and replaces it. This option defaults to ``0`` (no limit).

The following default configuration for a ``MongoClient`` works for most
applications:

.. code-block:: js
   
   const client = new MongoClient("<connection string>");

``MongoClient`` supports multiple concurrent requests. For each process,
create a client and reuse it for all operations in a process. This
practice is more efficient than creating a client for each request.

The driver does not limit the number of requests that
can wait for sockets to become available, and it is the application's
responsibility to limit the size of its pool to bound queuing
during a load spike. Requests wait for the amount of time specified in
the ``waitQueueTimeoutMS`` option, which defaults to ``0`` (no limit).

A request that waits more than the length of time defined by
``waitQueueTimeoutMS`` for a socket raises a connection error. Use this
option if it is more important to bound the duration of operations
during a load spike than it is to complete every operation.

When ``MongoClient.close()`` is called by any request, the driver
closes all idle sockets and closes all sockets that are in

> use as they are returned to the pool.

Reviewer: I would expand this sentence just a bit because it captures an important detail:
"closes all sockets that are in use **as they are returned** to the pool." Meaning we won't interrupt in progress operations. This has a unique downside for Node.js users that their process won't exit if say a slowQuery `find` is still pending results. 

Related issue https://jira.mongodb.org/browse/NODE-3263
We have no clean solution yet because there is risk to an API that cancels operations client side without making clear the impacts (think: unacknowledged write, without actually asking for it to be unacknowledged). There is a future where client-side operation timeout may provide an avenue for users to put a global cap on operation time, so that would provide an avenue for pending operations to be closed at least on _some_ kind of deadline.

TL;DR: I just think the "_as returned_" should be hoisted more in the phrasing b/c of its significance to our users.

.. _node-faq:

===
FAQ
===

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

This page contains frequently asked questions and their corresponding answers.

.. tip::

   If you can't find an answer to your problem on this page,
   see the :ref:`node-issues-help` page for next steps and more
   resources.

.. _node-faq-connection-pool:

How Does Connection Pooling Work in the Node Driver?
----------------------------------------------------

Every ``MongoClient`` instance has a built-in connection pool for each server
in your MongoDB topology. Connection pools open sockets on demand to
support concurrent requests to MongoDB in your application.

The maximum size of each connection pool is set by the ``maxPoolSize`` option, which
defaults to ``100``. If the number of in-use connections to a server reaches
the value of ``maxPoolSize``, the next request to that server will wait
until a connection becomes available.

Each ``MongoClient`` instance opens two additional sockets per server in your

> MongoDB topology for monitoring the server's state.

Author: This sentence is an introduction to the example in the next paragraph, which does relate to `maxPoolSize`. I think it should stay in this order

For example, a client connected to a 3-node replica set opens 6
monitoring sockets. It also opens as many sockets as needed to support
requests on each server, up to the value of ``maxPoolSize``. If
``maxPoolSize`` is ``100`` and the application only uses the primary
(the default), then only the primary connection pool grows and there can
be at most ``106`` total connections. If the application uses a

> :manual:`read preference </core/read-preference/#read-preference>` to

Author: I think "a read preference" is correct! there are multiple (optionally set) read preferences and a user can choose one of them for their operations.

query the secondary nodes, their pools also grow and there can be
``306`` total connections.

Additionally, connection pools are rate-limited such that each connection pool
can only create, at maximum, the value of ``maxConnecting`` connections
in parallel at any time. Any additional request stops waiting in the
following cases:

- The connection pool finishes creating a connection and there are fewer
  than ``minPoolSize`` connections in the pool.
- An existing connection is checked back into the pool.
- The driver's ability to reuse existing connections improves due to

>   rate-limits on connection creation.

Reviewer 2: [question]

Should this be "Any additional request stops waiting until one of the following cases"? 

Also - for the first case. If the connection pool just finished creating a connection, shouldn't that free up a slot? Because based on the previous wording, you're only talking about exceeding the ``maxConnecting`` parameter. And I think that if the connection pool never exceeded ``maxConnecting`` and just finished a connection, there would be an open spot.

Unless these three options is not meant to refer to exceeding ``maxConnecting``, in which case, they need to be placed elsewhere in the topic.

Author: To your first Q-- yes, it reads better with "until". changed the wording.

I'm a little confused at your second question -- my understanding is that there can be at max `maxConnecting` connections CREATED at the same time, and if your app wants to make another connection beyond that number, it will wait until a connection is freed up. 

I do think there is a typo and the first bullet point should read:
- The connection pool finishes creating a connection and there are fewer
  than ``maxPoolSize`` connections in the pool.

You can set the minimum number of concurrent connections to
each server with the ``minPoolSize`` option, which defaults to ``0``.

> The connection pool will be initialized with this number of sockets. If

Reviewer 2: [grammar]

Passive voice. 

```suggestion
The driver initializes the connection pool with this number of sockets. If
```

> sockets are closed due to any network errors, causing the total number

Reviewer 2: [nitpick]

Could they be closed for other reasons? Perhaps it might be best just to say they are closed or errors in general?

Author: Sure, I can just removed the "due to..."

of sockets (both in use and idle) to drop below the minimum, more
sockets are opened until the minimum is reached.

> You can set the maximum number of milliseconds that a connection can
> remain idle in the pool before being removed and replaced with
> the ``maxIdleTimeMS`` option, which defaults to ``0`` (no limit).

Reviewer 2: [suggestion]

Consider breaking this up. It exceeds the recommended sentence length and can be difficult to parse.

```suggestion
You can use ``maxIdleTimeMS`` to set the maximum number of milliseconds a connection can
remain idle in the pool. Once a connection has been idle for ``maxIdleTimeMS``, the driver removes and replaces it. ``maxIdleTimeMS`` defaults to ``0`` and has no upper limit.
```

The following default configuration for a ``MongoClient`` works for most
applications:

.. code-block:: js
   
   const client = new MongoClient("<connection string>");

Create a client once for each process, and reuse it for all
operations. It is a common mistake to create a new client for each
request, which is very inefficient.

To support high numbers of concurrent MongoDB requests
within one process, you can increase ``maxPoolSize``. Once the pool
reaches its maximum size, additional requests wait for sockets
to become available.

The driver does not limit the number of requests that

> can wait for sockets to become available and it is the application's

Reviewer 2: [nitpick]

Grammar

```suggestion
can wait for sockets to become available, and it is the application's
```
